src/Features/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticSources/DiagnosticSourceManager.cs

```diff
@@ -0,0 +1,133 @@
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics.DiagnosticSources;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.LanguageServer.Protocol;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics;
+
+[Export(typeof(IDiagnosticSourceManager)), Shared]
+internal class DiagnosticSourceManager : IDiagnosticSourceManager
+{
+    private readonly ImmutableDictionary<string, IDiagnosticSourceProvider> _documentProviders;
+    private readonly ImmutableDictionary<string, IDiagnosticSourceProvider> _workspaceProviders;
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public DiagnosticSourceManager([ImportMany] IEnumerable<IDiagnosticSourceProvider> sourceProviders)
+    {
+        _documentProviders = sourceProviders
+                .Where(p => p.IsDocument)
+                .ToImmutableDictionary(kvp => kvp.Name, kvp => kvp);
+
+        _workspaceProviders = sourceProviders
+                .Where(p => !p.IsDocument)
+                .ToImmutableDictionary(kvp => kvp.Name, kvp => kvp);
+    }
+
+    /// <inheritdoc />
+    public IEnumerable<string> GetSourceNames(bool isDocument)
+        => (isDocument ? _documentProviders : _workspaceProviders).Keys;
+
+    /// <inheritdoc />
+    public async ValueTask<ImmutableArray<IDiagnosticSource>> CreateDiagnosticSourcesAsync(RequestContext context, string? sourceName, bool isDocument, CancellationToken cancellationToken)
+    {
+        var providersDictionary = isDocument ? _documentProviders : _workspaceProviders;
+        if (sourceName != null)
+        {
+            // VS does not distinguish between document and workspace sources. Thus it can request
+            // document diagnostics with workspace source name. We need to handle this case.
+            if (providersDictionary.TryGetValue(sourceName, out var provider))
+                return await provider.CreateDiagnosticSourcesAsync(context, cancellationToken).ConfigureAwait(false);
+
+            context.TraceInformation($"No `{sourceName}` diagnostics for {isDocument}");
+            return [];
+        }
+        else
+        {
+            // VS Code (and legacy VS ?) pass null sourceName when requesting all sources.
+            using var _ = ArrayBuilder<IDiagnosticSource>.GetInstance(out var sourcesBuilder);
+            foreach (var kvp in providersDictionary)
+            {
+                // Exclude Task diagnostics from the aggregated sources.
+                if (kvp.Key != PullDiagnosticCategories.Task)
```

Comment:
Can we move the check for task diagnostics into the task diagnostic source provider itself?   Basically in the task diagnostic source provider, check if the HasVisualStudioLspCapability is set to true (otherwise don't return any sources)

Reply:
i strongly agree with this.  It seems odd that hte point of this PR was to not have the core system be aware of the individual diagnostic components, but this now hardcodes back in knowledge of Task diagnostics back in the core system.

Reply:
When I did that following tests failed. They had `useVSDiagnostics` set to `true`
```
TestDocumentTodoCommentsDiagnosticsForOpenFile_NoCategory
TestWorkspaceTodoAndDiagnosticForClosedFilesWithFSAOnAndTodoOn
TestWorkspaceTodoForClosedFilesWithFSAOffAndTodoOn
TestWorkspaceTodoForClosedFilesWithFSAOnAndTodoOn
```